src/main/java/de/tum/in/www1/artemis/web/rest/repository/RepositoryProgrammingExerciseParticipationResource.java

```diff
@@ -181,9 +183,14 @@ boolean canAccessRepository(Long participationId) throws IllegalArgumentExceptio
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<Map<String, String>> getFilesWithContent(@PathVariable Long participationId) {
         return super.executeAndCheckForExceptions(() -> {
-            // TODO: we need to fetch the commit which belongs to the last not ILLEGAL submission!
             Repository repository = getRepository(participationId, RepositoryActionType.READ, true);
-            var filesWithContent = super.repositoryService.getFilesWithContent(repository);
+
+            // Retrieve the last legal submission because we are getting the files for that commit.
+            var participation = participationService.findProgrammingExerciseParticipationWithLatestSubmissionAndResult(participationId);
+            var legalSubmissions = participation.getSubmissions().stream().filter(submission -> !submission.getType().equals(SubmissionType.ILLEGAL)).collect(Collectors.toList());
+            var latestLegalSubmission = (ProgrammingSubmission) legalSubmissions.get(legalSubmissions.size() - 1);
```

Comment:
there is a function which gets the latest submission for you. Participation.java: findLatestSubmission()